packages/wrangler/src/d1/execute.tsx

```diff
@@ -112,72 +116,79 @@ export async function executeSql(
 	}
 
 	return local
-		? await executeLocally(config, name, shouldPrompt, queries, persistTo)
-		: await executeRemotely(config, name, shouldPrompt, batchSplit(queries));
+		? await executeLocally(config, name, shouldPrompt, queries, persistTo, json)
+		: await executeRemotely(
+				config,
+				name,
+				shouldPrompt,
+				batchSplit(queries),
+				json
+		  );
 }
 type HandlerOptions = StrictYargsOptionsToInterface<typeof Options>;
-export const Handler = withConfig<HandlerOptions>(
-	async ({
+
+export const Handler = async (args: HandlerOptions): Promise<void> => {
+	const { local, database, yes, persistTo, file, command, json } = args;
+	if (json) {
+		// set loggerLevel to none to avoid readConfig warnings appearing in JSON output
+		logger.loggerLevel = "none";
+	}
+	const config = readConfig(args.config, args);
+	logger.log(d1BetaWarning);
+	if (file && command)
+		return logger.error(`Error: can't provide both --command and --file.`);
+
+	const isInteractive = process.stdout.isTTY;
+	const response: QueryResult[] | null = await executeSql(
+		local,
 		config,
 		database,
+		isInteractive && !yes,
+		persistTo,
 		file,
 		command,
-		local,
-		persistTo,
-		yes,
-	}): Promise<void> => {
-		logger.log(d1BetaWarning);
-		if (file && command)
-			return logger.error(`Error: can't provide both --command and --file.`);
-
-		const isInteractive = process.stdout.isTTY;
-		const response: QueryResult[] | null = await executeSql(
-			local,
-			config,
-			database,
-			isInteractive && !yes,
-			persistTo,
-			file,
-			command
-		);
+		json
+	);
 
-		// Early exit if prompt rejected
-		if (!response) return;
+	// Early exit if prompt rejected
+	if (!response) return;
 
-		if (isInteractive) {
-			// Render table if single result
-			render(
-				<Static items={response}>
-					{(result) => {
-						// batch results
-						if (!Array.isArray(result)) {
-							const { results, query } = result;
+	if (isInteractive && !json) {
+		// Render table if single result
+		render(
+			<Static items={response}>
+				{(result) => {
+					// batch results
+					if (!Array.isArray(result)) {
+						const { results, query } = result;
 
-							if (Array.isArray(results) && results.length > 0) {
-								const shortQuery = shorten(query, 48);
-								return (
-									<>
-										{shortQuery ? <Text dimColor>{shortQuery}</Text> : null}
-										<Table data={results}></Table>
-									</>
-								);
-							}
+						if (Array.isArray(results) && results.length > 0) {
+							const shortQuery = shorten(query, 48);
+							return (
+								<>
+									{shortQuery ? <Text dimColor>{shortQuery}</Text> : null}
+									<Table data={results}></Table>
+								</>
+							);
 						}
-					}}
-				</Static>
-			);
-		} else {
-			logger.log(JSON.stringify(response, null, 2));
-		}
+					}
+				}}
+			</Static>
+		);
+	} else {
+		// set loggerLevel back to log to actually output the JSON in stdout
+		logger.loggerLevel = "log";
```

Comment:
Again, what happens if the user had actually set log output level to, say, `error`. Here you are overriding that. I wonder if that is what you actually want?
In any case, I think we should then revert the level to whatever it was before in case this is part of a long-lived process.